keras/layers/recurrent.py

```diff
@@ -456,3 +457,150 @@ def get_config(self):
+    
+    # Examples:
+    ```python
+    model = Sequential()
+    model.add(BidirectionalRNN(LSTM(10, input_shape=(10, 20))))
+    model.add(Activation('softmax'))
+    
+    model.compile(loss='categorical_crossentropy', optimizer='rmsprop')
+
+    model.fit([X_train,], Y_train, batch_size=32, nb_epoch=20,
+              validation_data=([X_test], Y_test))
+    ```
+    '''
+    def __init__(self, rnn, merge_mode='concat', weights=None):
+
+        if K._BACKEND != 'theano':
+            warn('Bidirectional layers with input mask are fully supported only when using theano backend')
+
+        self.forward = rnn
+        self.reverse = deepcopy(rnn)
+        self.merge_mode = merge_mode
+        if weights:
+            nw = len(weights)
+            self.forward.initial_weights = weights[:nw/2]
+            self.reverse.initial_weights = weights[nw/2:]
+        self._cache_enabled = True
+
+    def get_weights(self):
+        return self.forward.get_weights() + self.reverse.get_weights()
+
+    def set_weights(self, weights):
+        nw = len(weights)
+        self.forward.set_weights(weights[:nw/2])
+        self.reverse.set_weights(weights[:nw/2])
+
+    def set_previous(self, layer):
+        self.previous = layer
+        self.forward.set_previous(layer)
+        self.reverse.set_previous(layer)
+
+    @property
+    def cache_enabled(self):
+        return self._cache_enabled
+
+    @cache_enabled.setter
+    def cache_enabled(self, value):
+        self._cache_enabled = value
+        self.forward.cache_enabled = value
+        self.reverse.cache_enabled = value
+
+    @property
+    def output_shape(self):
+        if self.merge_mode in ['sum', 'ave', 'mul']:
+            return self.forward.output_shape
+        elif self.merge_mode == 'concat':
+            shape = list(self.forward.output_shape)
+            shape[-1] *= 2
+            return tuple(shape)
+
+    def get_output(self, train=False):
+
+        X = self.get_input(train)
+        mask = self.get_input_mask(train)
+
+        X_rev = K.permute_dimensions(X, (1, 0, 2))
+        X_rev = X_rev[::-1]
+        X_rev = K.permute_dimensions(X_rev, (1, 0, 2))
+
+        mask_rev = mask
+
+        if mask:
+            mask_rev = K.permute_dimensions(mask_rev, (1, 0))
+            mask_rev = mask_rev[::-1]
+            mask_rev = K.permute_dimensions(mask_rev, (1, 0))
+
+            if K._BACKEND == 'theano':
+                #convert right padding to left padding
+                shifts = K.sum(mask_rev, axis=1)
+                import theano
+                X_rev, _ = theano.scan(lambda x, i: theano.tensor.roll(x, -i, 0),
+                             sequences=[X_rev, shifts])
+                mask_rev = mask
+
+        Y = self.forward(X, mask)
+        Y_rev = self.reverse(X_rev, mask_rev)
+        
+        if self.forward.return_sequences:
+            #Fix allignment
+            Y_rev = K.permute_dimensions((1, 0, 2))
+            Y_rev = [::-1]
```

Comment:
The output below still seems wrong to me, because you also have to shift the output to make the alignment right. 
Yes, that is why I said bidirectional is really ugly with padding in the first place.

[[[ 0.  6.]
  [ 1.  5.]
  [ 3.  3.]
  [ 6.  0.]]

 [[ 0.  3.]
  [ 0.  2.]
  [ 1.  0.]
  [ 3.  0.]]]